### PR TITLE
Fix an eta issue in elaborator-reflection

### LINKED ISSUE #1886: [ elab ] Partially applied function is not understood as a lambda in `>>=` of `Elab`
# Steps to Reproduce

Consider the code

```idris
import Language.Reflection      
                              
%default total                
                              
%language ElabReflection      
                     
x : List String      
x = ["a", "b"]                                       
                                                     
infixl 1 >==>                                         
      
-- From prelude:                                                     
--(>=>) : Monad m => (a -> m b) -> (b -> m c) -> (a -> m c)      
--(>=>) f g = \x => f x >>= g                                    
                                                                 
(>==>) : Monad m => (a -> m b) -> (b -> m c) -> (a -> m c)      
(>==>) f g = \x => f x >>= \r => g r                            
      
%runElab traverse_ (pure >==> logMsg "debug" 0) x -- works      
%runElab traverse_ (pure >=> \s => logMsg "debug" 0 s) x -- works      
%runElab traverse_ (pure >=> logMsg "debug" 0) x -- should work, but does not
```

and typecheck.

# Expected Behavior

```
LOG debug:0: a
LOG debug:0: b
LOG debug:0: a
LOG debug:0: b
LOG debug:0: a
LOG debug:0: b
```

# Observed Behavior

```
LOG debug:0: a
LOG debug:0: b
LOG debug:0: a
LOG debug:0: b
Error: Bad elaborator script Bind (pure "a") (logMsg "debug" 0) (non-function RHS of a Bind: $resolved2460 [2 closures]).

FishInElab:23:1--23:49
 19 | (>==>) f g = \x => f x >>= \r => g r
 20 |
 21 | %runElab traverse_ (pure >==> logMsg "debug" 0) x -- works
 22 | %runElab traverse_ (pure >=> \s => logMsg "debug" 0 s) x -- works
 23 | %runElab traverse_ (pure >=> logMsg "debug" 0) x -- should work, but does not
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


```

### PR DESCRIPTION
**Pull request by Russoul**
Fixes #1886

What's interesting is that the code snippet provided in the linked issue works against a sufficiently recent Idris build.

I added another test-case that fails against recent Idris builds.

@buzden feel free to throw in some more test-cases, so we have an extensive collection. Just in case :)

### CODE REVIEW

File: src/TTImp/Elab/RunElab.idr [hunk 1]
```diff
@@ -88,7 +88,12 @@ elabScript fc nest env script@(NDCon nfc nm t ar args) exp
                       elabScript fc nest env
                               !(sc defs (toClosure withAll env
                                               !(quote defs env act'))) exp
-                  x => failWith defs $ "non-function RHS of a Bind: " ++ show x
+                  head => do -- we've got an eta-contracted form here. Eta-expand!
```
Comment (gallais) [inline src/TTImp/Elab/RunElab.idr:91]: This comment needs clarifying: we're not eta-expanding anything here, we're forming `k act'` (where `act'`
is `act` with `Elab` stripped off by `elabScript` I hope?)

Comment (gallais) [inline src/TTImp/Elab/RunElab.idr:91]: This comment needs clarifying: we're not eta-expanding anything here, we're forming `k act'` (where `act'`
is `act` with `Elab` stripped off by `elabScript` I hope?)

Comment (Russoul) [reply src/TTImp/Elab/RunElab.idr]: Indeed we are not eta-expanding...
Restructured the logic around this recipe:
1) run elabScript on `act` effectively stripping off `Elab` from the type
2) apply k to (1)
3) run elabScript on the result

**Review by gallais — APPROVED** [review]

**Comment by buzden** [comment]
By the way, I realised that the original bug was found in the value-returning mode of `%runElab`. The mode usage was changed due to #1878. Since quotation seems to be present in this PR's computation, maybe it's worth to add alternative variants to the tests, like
```idris
x1 : Unit
x1 = %runElab traverse_ ...
```

However, the old ones are working in this mode too.

**Comment by Russoul** [comment]
@ohad @gallais Any idea what could be causing Frex failure?

**Comment by gallais** [comment]
There's a suspicious search failure on the windows CI. I'll try to merge & rerun the suite.

File: src/TTImp/Elab/RunElab.idr [hunk 2]
```diff
@@ -81,14 +81,17 @@ elabScript fc nest env script@(NDCon nfc nm t ar args) exp
...
              evalClosure empty val
     elabCon defs "Bind" [_,_,act,k]
-        = do act' <- elabScript fc nest env
+        -- act : Elab A
+        -- k : A -> Elab B
+        -- 1) Run elabScript on act stripping off Elab
+        -- 2) apply k to the result of (1)
+        -- 3) Run elabScript on the result stripping off Elab
+        = do act <- elabScript fc nest env
                                 !(evalClosure defs act) exp
```
Comment (gallais) [inline src/TTImp/Elab/RunElab.idr:90]: This `exp` is suspicious: it's the type of the overall elab script, not of `act`.
```suggestion
                                !(evalClosure defs act) Nothing
```

**Comment by Russoul** [comment]
Hm, okay, I guess we'll need to extract a proper type from `Bind`.

**Comment by gallais** [comment]
Or is the `Maybe (Glued vars)` the type of the hole? In which case passing `exp` along was correct. :thinking: 

https://github.com/idris-lang/Idris2/blob/1d24a4d5043fe93b0bd9aaa4281413cd9c059d30/src/TTImp/Elab/RunElab.idr#L166-L171

**Comment by Russoul** [comment]
Yep, based on the snippet in your comment that seems to be true.

**Comment by gallais** [comment]
So the windows build is now once more working. Frex is the only thing left.

**Comment by buzden** [comment]
> So the windows build is now once more working. Frex is the only thing left.

In my [branch with the additional test](https://github.com/buzden/Idris2/tree/reflection-bind-eta) frex test was the only failing... :thinking:

**Comment by gallais** [comment]
Okay I'm hitting this bug in my work on deriving `Functor`.
So I think I'll find the courage to dig into the frex error.

**Comment by Russoul** [comment]
Let's see if it works now

**Comment by Russoul** [comment]
Don't merge yet please. Going to add another test (due to @buzden)

**Comment by Russoul** [comment]
@gallais All done

**Pull request merged**